src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java

```diff
@@ -98,37 +173,45 @@ public static Predicate buildRoleAssignmentConstraints(
-            .stream().filter(RoleAssignmentFilter::filterByActiveRole).collect(Collectors.toList());
-
-        RoleAssignmentSearchData searchData = new RoleAssignmentSearchData(
-            activeRoleAssignments,
-            RoleAssignmentForSearch::getRoleType
-        );
-        List<RoleAssignmentForSearch> roleAssignmentsForSearch = searchData.getRoleAssignmentsForSearch();
+    private static Predicate getPermissionRequirementPredicate(PermissionRequirements permissionsRequired,
+                                                               CriteriaBuilder builder,
+                                                               Join<TaskResource, TaskRoleResource> taskRoleResources) {
+        PermissionRequirements nextRequirements = permissionsRequired;
+        PermissionJoin nextPermissionJoin = nextRequirements.getPermissionJoin();
+        Predicate permissionRequirementPredicate = null;
 
-        final Join<TaskResource, TaskRoleResource> taskRoleResources = root.join(TASK_ROLE_RESOURCES);
+        List<Predicate> permissionPredicates = new ArrayList<>();
+        Predicate[] predicates = new Predicate[0];
 
-        List<Predicate> rolePredicates = new ArrayList<>();
-        for (RoleAssignmentForSearch roleAssignment : roleAssignmentsForSearch) {
+        while (nextRequirements != null && !nextRequirements.isEmpty()) {
+            PermissionRequirement requirement = nextRequirements.getPermissionRequirement();
 
-            Predicate rolePredicate = builder.equal(
-                taskRoleResources.get(ROLE_NAME_COLUMN),
-                roleAssignment.getRoleName()
-            );
-
-            rolePredicates.add(builder.and(rolePredicate, builder.isTrue(taskRoleResources.get(READ_COLUMN))));
+            for (PermissionTypes type : requirement.getPermissionTypes()) {
+                permissionPredicates.add(builder.isTrue(taskRoleResources.get(type.value().toLowerCase(Locale.ROOT))));
+            }
+            Predicate permissionPredicate;
+            if (PermissionJoin.AND.equals(requirement.getPermissionJoin())) {
+                permissionPredicate = builder.and(permissionPredicates.toArray(predicates));
+            } else {
+                permissionPredicate = builder.or(permissionPredicates.toArray(predicates));
+            }
+            permissionPredicates.clear();
+
+            if (permissionRequirementPredicate == null) {
+                permissionRequirementPredicate = permissionPredicate;
+            } else if (PermissionJoin.AND.equals(nextPermissionJoin)) {
+                permissionRequirementPredicate = builder.and(permissionRequirementPredicate, permissionPredicate);
+            } else {
+                permissionRequirementPredicate = builder.or(permissionRequirementPredicate, permissionPredicate);
+            }
 
+            nextPermissionJoin = nextRequirements.getPermissionJoin();
+            nextRequirements = nextRequirements.getNextPermissionRequirements();
```

Comment:
Think I get what you're doing here with current/next etc.

Reply:
Yes, here it loop through the collection of permission requirement.